common/autoware_lanelet2_utils/src/geometry.cpp

```diff
@@ -0,0 +1,221 @@
+{
+  std::vector<lanelet::Point3d> points;
+  const auto linestring = lanelet.centerline();
+  return interpolate_linestring(linestring, distance);
+}
+
+std::optional<lanelet::ConstPoint3d> interpolate_lanelet_sequence(
+  const lanelet::ConstLanelets & lanelet_sequence, double distance)
+{
+  const auto merged_sequence = lanelet::LaneletSequence(lanelet_sequence);
+  const auto centerline = merged_sequence.centerline();
+  return interpolate_linestring(centerline, distance);
+  ;
+}
+
+std::optional<lanelet::CompoundLineString3d> concatenate_center_line(
+  const lanelet::ConstLanelets & lanelets)
+{
+  if (lanelets.empty()) {
+    return std::nullopt;
+  }
+  const auto merged_sequence = lanelet::LaneletSequence(lanelets);
+  return merged_sequence.centerline();
+}
+
+std::optional<lanelet::LineString3d> get_linestring_from_arc_length(
+  const lanelet::ConstLineString3d & linestring, const double s1, const double s2)
+{
+  lanelet::Points3d points;
+  double accumulated_length = 0;
+  size_t start_index = linestring.size();
+  if (linestring.size() < 2) {
+    return std::nullopt;
+  }
+
+  const double total_length = lanelet::geometry::length(linestring);
+  if (s1 < 0.0 || s2 > total_length || s1 >= s2) {
+    return std::nullopt;
+  }
+
+  for (size_t i = 0; i < linestring.size() - 1; i++) {
+    const auto & p1 = linestring[i];
+    const auto & p2 = linestring[i + 1];
+    const double length = lanelet::geometry::distance3d(p1, p2);
+
+    if (accumulated_length + length > s1) {
+      start_index = i;
+      break;
+    }
+    accumulated_length += length;
+  }
+
+  if (start_index < linestring.size() - 1) {
+    const auto & p1 = linestring[start_index];
+    const auto & p2 = linestring[start_index + 1];
+    const double residue = s1 - accumulated_length;
+
+    const auto start_point = interpolate_point(p1, p2, residue);
+    if (!start_point.has_value()) return std::nullopt;
+    points.emplace_back(start_point.value());
+  }
+
+  size_t end_index = linestring.size();
+  for (size_t i = start_index; i < linestring.size() - 1; i++) {
+    const auto & p1 = linestring[i];
+    const auto & p2 = linestring[i + 1];
+    const double length = lanelet::geometry::distance3d(p1, p2);
+    if (accumulated_length + length > s2) {
+      end_index = i;
+      break;
+    }
+    accumulated_length += length;
+  }
+
+  for (size_t i = start_index + 1; i < end_index; i++) {
+    const auto p = lanelet::Point3d(linestring[i]);
+    points.emplace_back(p);
+  }
+  if (end_index < linestring.size() - 1) {
+    const auto & p1 = linestring[end_index];
+    const auto & p2 = linestring[end_index + 1];
+    const double residue = s2 - accumulated_length;
+    const auto end_point = interpolate_point(p1, p2, residue);
+    points.emplace_back(lanelet::Point3d(linestring[end_index]));
+
+    if (!end_point.has_value()) return std::nullopt;
+
+    points.emplace_back(lanelet::Point3d(lanelet::InvalId, end_point.value()));
```

Comment:
It looks like there’s no real benefit to using `emplace_back` here.

```suggestion
    points.emplace_back(lanelet::InvalId, end_point.value());
```